src/build.sh

```diff
@@ -1,21 +1,199 @@
 #!/usr/bin/env bash
 set -eu
 
+#===============================================================================
+# Assuming this script is run with: `cd $GERBIL_BASE/src && ./build.sh`
+#===============================================================================
+
+## global setup
+readonly GERBIL_SOURCE=$(pwd -P)
+readonly GERBIL_BASE="$(dirname "${GERBIL_SOURCE}")"
+readonly BUILD_DIR="${GERBIL_SOURCE}/build"
+
+#===============================================================================
+## feedback
+feedback_err() {
+  (1>&2 echo "*** ERROR; ${1}")
+}
+
+feedback_low() {
+  echo "[*] ${1}"
+}
+
+feedback_mid() {
+  echo ">>> ${1}"
+}
+
 die() {
-    echo "*** ERROR; build failed"
-    exit 1
+  feedback_err "build failed"
+  exit 1
 }
 
-build_gerbil() {
-    ./build0.sh || die
-    ./build1.sh final || die
-    ./build2_fini.sh
-    ./build_stdlib.sh || die
-    ./build_lang.sh || die
-    ./build_tools.sh || die
+## bootstrap
+target_setup () {
+  local target="${1}"
+  # make sure 'target' expands to something: don't want to 'rm -rf /bin'
+  rm -rf "${target:?}/"{bin,lib} 
+  mkdir -p "${target}/"{bin,lib}
+}
+
+compile_runtime () {
+  local target_lib="${1}"
+  (cd gerbil/runtime && ./build.scm "${target_lib}")
+}
+
+finalize_build () {
+  local target_lib="${1}"
+  local target_bin="${2}"
+  cp -v gerbil/boot/*.scm "${target_lib}"
+  cp -v gerbil/interactive/*.ss "${target_lib}"
+  cp -v gerbil/gxi gerbil/gxc gerbil/gxi-build-script "${target_bin}"
+  (cd "${target_bin}" && ln -s gxi gxi-script)
 }
 
-echo "Building Gerbil"
-build_gerbil
+stage0 () {
+  local gerbil_target="${GERBIL_BASE}/bootstrap/stage0"
+  local target_bin="${gerbil_target}/bin"
+  local target_lib="${gerbil_target}/lib"
+  
+  ## feedback
+  feedback_low "Building gerbil stage0"
+  
+  ## preparing target directory
+  feedback_mid "preparing ${gerbil_target}"
+  target_setup "${gerbil_target}"
+  
+  ## gerbil runtime
+  feedback_mid "compiling runtime"
+  compile_runtime "${target_lib}"
+  
+  ## gerbil bootstrap
+  feedback_mid "preparing bootstrap"
+  rsync -auv bootstrap/gerbil "${target_lib}"
+  find "${target_lib}" -name \*.scm > .build.stage0
+  
+  feedback_mid "compiling gerbil core"
+  gsi "${BUILD_DIR}/build0.scm" || die
+  
+  ## cleaning up
+  rm -f .build.stage0
+  
+  ## finalize build
+  feedback_mid "finalizing build"
+  finalize_build "${target_lib}" "${target_bin}"
+}
+
+stage1 () {
+  ## handling script arguments
+  readonly final_string="[final]"
+  if [[ "xfinal" = "x${1:-}" ]]; then
+      # if final stage, install directly to 'GERBIL_BASE'
+      local gerbil_target="${GERBIL_BASE}"
+      local final="1"
+  else
+      local gerbil_target="${GERBIL_BASE}/bootstrap/stage1"
```

Comment:
final is not so useful for building in stage1, but it's useful for keeping the bootstrap build.
Can we change its meaning to mean "build in target, but don't delete stage0" instead of building in stage1?